### ISSUE TITLE
MNIST For ML Beginners

### ISSUE THREAD

**Issue by killelder** [killelder/tensorflownote#3]
https://www.tensorflow.org/get_started/mnist/beginners

---------------------------------------------------------------
本來不是很想看這篇的, 不過還是決定來看看有沒有甚麼值得學習的
算是tensorflow的"hello world"

MNIST是手寫的dataset for machine learning
是一個Softmax Regression的model

整個sample code寫在  https://github.com/tensorflow/tensorflow/blob/r1.4/tensorflow/examples/tutorials/mnist/mnist_softmax.py

(這個範例google沒有使用hidden layer, 直接把input跟output層連接而已)
利用MNIST的資料, 需要先輸入下面兩行

from tensorflow.examples.tutorials.mnist import input_data
mnist = input_data.read_data_sets("MNIST_data/", one_hot=True)

會自己download MNIST的資料, MNIST分成三個部分
55000 training data, 10000 test data, 5000 validation data
這告訴了我們必須要測試我們從未訓練過的data,

MNIST包含了image 的pixel "data"和對應的答案"label"
暫且把images 當成"x", labels當成"y",
每一個images包含28x28=784的pixel, 
我們怎麼把images flattern(拉平, 就是變成1維的data) 並不重要, 只要每一個data都做同樣的flattern就好
每一個tensor是pixel的強度 0~1
當調用 mnist.train.images 他是一個tensor with shape of [55000, 784]
mnist.train.labels 是一個tensor with shape of [55000, 10]
我們打算把label換成"one-hot vectors"
one-hot vectors是把答案的n種可能性, 變成一個n dimension的向量, 其中對應到的那個答案為1
其他為0
例如 答案從0~9, 當答案為3的時候, one-hot vectors的形式為[0,0,0,1,0,0,0,0,0,0]

如果想看機率, softmax就是拿來做這件事的
因為softmax給我們一連串的機率介於0~1, 並且加起來=1
softmax可以想像是兩步驟,
1. 先加總所有的evidence
2. 把加總起來的evidence轉化為機率

(Google此時的圖我覺得蠻有趣的, 他把weights加起來, 因為如果負的weights表示他機率越小, 
正的weights表示他機率越大,可以從圖形的weights看出機器怎麼判斷的)

evidence = sigma(Wij ‧ xj + bi)
我們還有另外一個evidence 叫做bias,
i是10種答案的第幾種, j是pixel的位置
然後利用softmax求機率
y = softmax(evidence)

此時的softmax算是一個"activation" 或是 "link"的腳色, 把output變成我們想要的形式,
softmax = exp(evidence)/sigma(exp(evidence))
跟機率的概念一樣, 是拿某個/全部
但是做這個的用意更大是為了增加當算對的機率 
(因為exp變化一點點就差很多, 所以當某一項比較大的時候, 其他的機率會變的小很多, 會變成接近0)
softmax最後/sigma是為了讓他們加起來=1, 變得像一個機率分布

所以整個式子到這裡可以表示成
y = softmax(Wx + b)

----------------------------------------------------------
Implementing the Regression
接下來才真的要把tensorflow加進來完成上面那行式子的model

這一整段我不是很了解, 歡迎高手指教
[
為了做有效的運算in python, 我們通常使用Numpy做運算, 例如矩陣相乘,
在其他程式做運算的時候常常需要回傳到python, 
當使用GPUs或者分散式的方法的時候, 轉換data還是會造成很大的cost
Tensorflow也會把一些運算拿到python外面去做
]

未完......
